fedimint-client/src/sm/executor.rs

```diff
@@ -0,0 +1,614 @@
+use std::collections::BTreeMap;
+use std::sync::Arc;
+use std::time::{Duration, SystemTime};
+
+use anyhow::bail;
+use fedimint_core::core::{IntoDynInstance, ModuleInstanceId};
+use fedimint_core::db::{AutocommitError, Database};
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_core::task::TaskGroup;
+use fedimint_core::{impl_db_lookup, impl_db_record};
+use futures::future::select_all;
+use futures::stream::StreamExt;
+use tokio::select;
+use tracing::{debug, error, info, warn};
+
+use crate::sm::state::{DynContext, DynState};
+use crate::sm::{GlobalContext, State, StateTransition};
+use crate::ClientModule;
+
+/// After how many retries a DB transaction is aborted with an error
+const MAX_DB_RETRIES: Option<usize> = Some(100);
+
+/// Wait time till checking the DB for new state machines when there are no
+/// active ones
+const EXECUTOR_POLL_INTERVAL: Duration = Duration::from_secs(1);
+
+/// Prefixes for executor DB entries
+enum ExecutorDbPrefixes {
+    /// See [`ActiveStateKey`]
+    ActiveStates = 0xa1,
+    /// See [`InactiveStateKey`]
+    InactiveStates = 0xa2,
+}
+
+/// Executor that drives forward state machines under its management.
+///
+/// Each state transition is atomic and supposed to be idempotent such that a
+/// stop/crash of the executor at any point can be recovered from on restart.
+/// The executor is aware of the concept of Fedimint modules and can give state
+/// machines a different [execution context](super::state::Context) depending on
+/// the owning module, making it very flexible.
+pub struct Executor {
+    inner: Arc<ExecutorInner>,
+}
+
+struct ExecutorInner {
+    db: Database,
+    context: GlobalContext,
+    module_contexts: BTreeMap<ModuleInstanceId, DynContext>,
```

Comment:
We use a map to allow the executor to access context of different module instance than it's own?

Specifically wondering how "making change" will work in a generalized way that doesn't depend on some specific module that can do this -- e.g. our ecash module at X git hash.

Reply:
The executor is an executor for all modules. Depending with which module's state machine it is dealing it will select the corresponding context. The executor also isn't a full client yet, just the core piece.